File: quizJavascript/lib/DLList.js
'use strict';

const DNode = require('./DNode.js');

module.exports = (() => {

    let head, tail, size;

    function DLList(data) {
        _initDDList();

        if (null != data && data.length != 0) {
            data.forEach(ite => this.addLast(ite));
        }
    }

    function _initDDList() {
        head = new DNode();
        tail = new DNode();

        size = 0;

        head.next = tail;
        tail.previous = head;
    }

    function _addBefore(referenceNode, element) {
        const prevNode = _getPrev(referenceNode);
        const newNode = new DNode(element, prevNode, referenceNode);

        referenceNode.previous = newNode;
        prevNode.next = newNode;

        size++;
    }

    function _addAfter(referenceNode, element) {
        const nextNode = _getNext(referenceNode);
        const newNode = new DNode(element, referenceNode, nextNode);

        referenceNode.next = newNode;
        nextNode.previous = newNode;

        size++;
    }

    function _getNext(referenceNode) {
        if (referenceNode == tail)
            throw new Error("Reference node cannot be equal to the tail");

        return referenceNode.next;
    }

    function _getPrev(referenceNode) {
        if (referenceNode == head)
            throw new Error("Reference node cannot be equal to the head");

        return referenceNode.previous;
    }

    function _remove(removeNode) {
        const prevNode = _getPrev(removeNode);
        const nextNode = _getNext(removeNode);

        removeNode.previous = null;
        removeNode.next = null;

        prevNode.next = nextNode;
        nextNode.previous = prevNode;

        size--;

        return removeNode.getData();
    }

    function _removeCheckEmptyList() {
        if (this.isEmpty())
            throw new Error("Double Linked list is empty, cannot remove element");
    }

    DLList.prototype = {
        isEmpty() {
            return (size === 0);
        },

        getSize() {
            return size;
        },

        clear() {
            head = null;
            tail = null;

            size = 0;
        },

        addFirst(element) {
            _addAfter(head, element);
        },

        addLast(element) {
            _addBefore(tail, element);
        },

        removeLast() {
            _removeCheckEmptyList.call(this);

            return _remove(_getPrev(tail));
        },

        removeFirst() {
            _removeCheckEmptyList.call(this);

            return _remove(_getNext(head));
        },

        toString() {
            let strTabs = [];

            if (null != head) {
                let probe = head.next;

                while (probe != tail) {
                    strTabs.push(probe.toString());
                    probe = probe.next;
                }
            }

            return "[" + strTabs.join(" ") + "]";
        }
    };

    return DLList;
})();

File: quizDotNet/QuizDLList/DLList.cs
﻿using System.Collections.Generic;
using System.Text;
using QuizDLList.Exception;

namespace QuizDLList
{

    public class DlList<E>
    {
        private DNode<E> Head { get; set; }
        private DNode<E> Tail { get; set; }

        public int Size { get; private set; }

        public DlList(params E[] es)
        {
            InitDdList();

            if (null == es || es.Length == 0) return;

            foreach (var e in es)
            {
                AddLast(e);
            }
        }

        private void InitDdList()
        {
            Head = new DNode<E>();
            Tail = new DNode<E>();

            Size = 0;

            Head.Next = Tail;
            Tail.Previous = Head;
        }

        private DNode<E> GetNext(DNode<E> referenceNode)
        {
            if (referenceNode == Tail)
                throw new BoundaryViolationException("Reference node cannot be equal to the tail");

            return referenceNode.Next;
        }

        private DNode<E> GetPrev(DNode<E> referenceNode)
        {
            if (referenceNode == Head)
                throw new BoundaryViolationException("Reference node cannot be equal to the head");

            return referenceNode.Previous;
        }

        private void AddBefore(DNode<E> referenceNode, E element)
        {
            var prevNode = this.GetPrev(referenceNode);
            var newNode = new DNode<E>(element, prevNode, referenceNode);

            referenceNode.Previous = newNode;
            prevNode.Next = newNode;

            Size++;
        }

        private void AddAfter(DNode<E> referenceNode, E element)
        {
            var nextNode = this.GetNext(referenceNode);
            var newNode = new DNode<E>(element, referenceNode, nextNode);

            referenceNode.Next = newNode;
            nextNode.Previous = newNode;

            Size++;
        }

        public void AddFirst(E element)
        {
            AddAfter(Head, element);
        }

        public void AddLast(E element)
        {
            AddBefore(Tail, element);
        }

        private E Remove(DNode<E> removeNode)
        {
            var prevNode = this.GetPrev(removeNode);
            var nextNode = this.GetNext(removeNode);

            removeNode.Previous = null;
            removeNode.Next = null;

            prevNode.Next = nextNode;
            nextNode.Previous = prevNode;

            Size--;

            return removeNode.Element;
        }

        public E RemoveLast()
        {
            RemoveCheckEmptyList();

            return this.Remove(this.GetPrev(Tail));
        }

        public E RemoveFirst()
        {
            RemoveCheckEmptyList();

            return this.Remove(this.GetNext(Head));
        }

        private void RemoveCheckEmptyList()
        {
            if (this.IsEmpty())
                throw new EmptyListException("Double Linked list is empty, cannot remove element");
        }

        public void Clear()
        {
            this.InitDdList();
        }

        public bool IsEmpty()
        {
            return Size == 0;
        }

        public override string ToString()
        {
            var str = new List<string>();

            if (null != Head)
            {
                var probe = Head.Next;

                while (probe != Tail)
                {
                    str.Add(probe.ToString());
                    probe = probe.Next;
                }
            }

            var builder = new StringBuilder();
            builder.Append("[");
            builder.Append(string.Join(" ", str));
            builder.Append("]");
            return builder.ToString();
        }
    }

}
